src/puppetlabs/puppetdb/query_eng/engine.clj

```diff
@@ -1795,34 +1813,89 @@
+  (let [parts (parse/parse-field field)]
+    (if (some #(= ::parse/match-field-part (:kind %)) parts)
+      (let [[head & path] parts
+            path (if (= (:name head) "trusted")
+                   ;; Real location is facts.trusted...
+                   (cons head path)
+                   path)
+            ;; Q: factpath-regexp-to-regexp?
+            pg-rx (->> path
+                       (mapcat #(case (:kind %)
+                                  ::parse/indexed-field-part [(:name %) (:index %)]
+                                  ::parse/match-field-part [(:pattern %)]
+                                  ::parse/named-field-part [(:name %)]))
+                       (string/join "#~"))
+            ;; Right now, this code communicates the paths via AST
+            ;; dotted fields, which is insufficient because the
+            ;; current syntax can't represent all possible fact names
+            ;; (see db-path->ast-field).
+            ;;
+            ;; Eventually, and depending on what we decide about
+            ;; match(), we might want to either:
+            ;;   - move this down into the ->plan functions so we
+            ;;     don't have to round-trip through AST, or otherwise
+            ;;     allow us to carry the parsed field representation
+            ;;     rather than a reconsituted AST field string.
+            ;;   - rework AST's quoting syntax (in a new endpoint
+            ;;     version)
+            ;;   - just rely on the well specified syntax we already
+            ;;     have, and provide support paths as JSON vectors,
+            ;;     e.g. ["foo" "bar"], ["foo" ["nth" "bar" 5], "baz"],
+            ;;     ["foo" ["match" "\\d+"]], etc.
+            fact_paths (->> (jdbc/query-to-vec (str "SELECT path_array FROM fact_paths"
+                                                    "  WHERE (path ~ ? AND path IS NOT NULL)")
+                                               (doto (PGobject.)
+                                                 (.setType "text")
+                                                 (.setValue pg-rx)))
+                            (map :path_array)
+                            (map (fn [path]
+                                   (if (= (first path) "trusted")
+                                     path
+                                     (cons "facts" path))))
+                            ;; cf. parse-dot-query which relies on the
+                            ;; fact that any array indexing will match
+                            ;; \d+, i.e. foo.5.bar.
+                            (map db-path->ast-field))]
+        (case (count fact_paths)
+          0 ["or" "false"] ;; Could this just be false?
+          1 (vector operator (first fact_paths) value)
+          (into ["or"] (map #(vector operator % value) fact_paths))))
+      (if (= "trusted" (::parse/named-field-part (first parts)))
+        [operator (str "facts." field) value]
+        [operator field value]))))
```

Comment:
added this `if` statement so that we don't lose the `trusted` facts optimization